File: backend-springboot/src/main/java/com/cosmetics/model/Coupon.java
package com.cosmetics.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.time.LocalDateTime;

/**
 * Model mã giảm giá — TV3
 *
 * discountType:
 *   "percent" → giảm theo %, có thể có maxDiscountAmount
 *   "fixed"   → giảm số tiền cố định
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "coupons")
public class Coupon {

    @Id
    private String id;

    /** Mã coupon (viết hoa, unique) — VD: SUMMER20 */
    @Indexed(unique = true)
    private String code;

    /** Mô tả ngắn hiển thị cho khách */
    private String description;

    /**
     * Loại giảm giá: "percent" | "fixed"
     */
    @Field("discount_type")
    @Builder.Default
    private String discountType = "percent";

    /**
     * Giá trị giảm:
     *   - Nếu percent: 0-100 (%)
     *   - Nếu fixed: số tiền VND
     */
    private Double value;

    /** Giá trị đơn hàng tối thiểu để áp dụng (0 = không giới hạn) */
    @Field("min_order_amount")
    @Builder.Default
    private Double minOrderAmount = 0.0;

    /** Giảm tối đa (chỉ áp dụng cho percent, null = không giới hạn) */
    @Field("max_discount_amount")
    private Double maxDiscountAmount;

    /** Số lần sử dụng tối đa (null = không giới hạn) */
    @Field("usage_limit")
    private Integer usageLimit;

    /** Số lần đã sử dụng */
    @Field("used_count")
    @Builder.Default
    private Integer usedCount = 0;

    /** Ngày hết hạn (null = không hết hạn) */
    @Field("expires_at")
    private LocalDateTime expiresAt;

    /** Kích hoạt / vô hiệu hóa */
    @Field("is_active")
    @Builder.Default
    private Boolean isActive = true;

    @CreatedDate
    @Field("created_at")
    private LocalDateTime createdAt;
}

File: backend-nodejs/src/models/Coupon.js
import mongoose from 'mongoose';

/**
 * Model mã giảm giá — TV3
 *
 * discountType:
 *   "percent" → giảm theo %, có thể có max_discount_amount
 *   "fixed"   → giảm số tiền cố định
 */
const couponSchema = new mongoose.Schema(
  {
    code: {
      type: String,
      required: true,
      unique: true,
      uppercase: true,
      trim: true,
    },
    description: { type: String, default: '' },

    // "percent" | "fixed"
    discount_type: {
      type: String,
      enum: ['percent', 'fixed'],
      default: 'percent',
    },

    // Giá trị giảm: nếu percent → 0-100, nếu fixed → số tiền VND
    value: { type: Number, required: true },

    // Giá trị đơn tối thiểu để áp dụng (0 = không giới hạn)
    min_order_amount: { type: Number, default: 0 },

    // Giảm tối đa (chỉ dùng cho percent, null = không giới hạn)
    max_discount_amount: { type: Number, default: null },

    // Số lần sử dụng tối đa (null = không giới hạn)
    usage_limit: { type: Number, default: null },

    // Số lần đã sử dụng
    used_count: { type: Number, default: 0 },

    // Ngày hết hạn (null = không hết hạn)
    expires_at: { type: Date, default: null },

    // Kích hoạt / vô hiệu hóa
    is_active: { type: Boolean, default: true },
  },
  { timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' } }
);

// Map dữ liệu đầu ra cho frontend
couponSchema.set('toJSON', {
  transform: (doc, ret) => {
    ret.id = ret._id;
    ret.discountType = ret.discount_type;
    ret.minOrderAmount = ret.min_order_amount;
    ret.maxDiscountAmount = ret.max_discount_amount;
    ret.usageLimit = ret.usage_limit;
    ret.usedCount = ret.used_count;
    ret.expiresAt = ret.expires_at;
    ret.isActive = ret.is_active;
    ret.createdAt = ret.created_at;

    delete ret._id;
    delete ret.__v;
    delete ret.discount_type;
    delete ret.min_order_amount;
    delete ret.max_discount_amount;
    delete ret.usage_limit;
    delete ret.used_count;
    delete ret.expires_at;
    delete ret.is_active;
    delete ret.created_at;
    delete ret.updated_at;

    return ret;
  },
});

export default mongoose.model('Coupon', couponSchema);
